src/prof_recent.c

```diff
@@ -0,0 +1,407 @@
+#define JEMALLOC_PROF_C_
+#include "jemalloc/internal/jemalloc_preamble.h"
+#include "jemalloc/internal/jemalloc_internal_includes.h"
+
+#include "jemalloc/internal/assert.h"
+#include "jemalloc/internal/emitter.h"
+
+ssize_t opt_prof_recent_alloc_max = PROF_RECENT_ALLOC_MAX_DEFAULT;
+static malloc_mutex_t prof_recent_alloc_mtx; /* Protects the fields below */
+static ssize_t prof_recent_alloc_max;
+static ssize_t prof_recent_alloc_count = 0;
+static prof_recent_t *prof_recent_alloc_dummy = NULL;
+
+static const nstime_t zero_time = NSTIME_ZERO_INITIALIZER;
+
+void
+prof_recent_alloc_prepare(tsd_t *tsd, prof_tctx_t *tctx) {
+	assert(opt_prof && prof_booted);
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), tctx->tdata->lock);
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	/*
+	 * Increment recent_count to hold the tctx so that it won't be gone
+	 * even after tctx->tdata->lock is released.  This acts as a
+	 * "placeholder"; the real recording of the allocation requires a lock
+	 * on prof_recent_alloc_mtx and is done in prof_recent_alloc_slow (when
+	 * tctx->tdata->lock has been released).
+	 */
+	++tctx->recent_count;
+}
+
+static void
+decrement_recent_count(tsd_t *tsd, prof_tctx_t *tctx) {
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	if (tctx == NULL) {
+		return;
+	}
+	malloc_mutex_lock(tsd_tsdn(tsd), tctx->tdata->lock);
+	assert(tctx->recent_count > 0);
+	--tctx->recent_count;
+	prof_tctx_try_destroy(tsd, tctx);
+}
+
+#ifndef JEMALLOC_JET
+static
+#endif
+prof_recent_t *
+extent_prof_recent_alloc_get(const extent_t *extent) {
+	malloc_mutex_assert_owner(TSDN_NULL, &prof_recent_alloc_mtx);
```

Comment:
Sadly with TSDN_NULL this doesn't check anything -- https://github.com/jemalloc/jemalloc/blob/4b76c684bb8d7f0b7960bfac84391e9fd51a234e/include/jemalloc/internal/witness.h#L190

Let's pass tsd down.